pgxn/neon/pagestore_smgr.c

```diff
@@ -332,16 +335,22 @@ compact_prefetch_buffers(void)
 		/* Adjust the location of our known-empty slot */
 		empty_ring_index--;
 
+		/* empty the moved slot */
 		source_slot->status = PRFS_UNUSED;
 		source_slot->buftag = (BufferTag) {0};
 		source_slot->response = NULL;
 		source_slot->my_ring_index = 0;
 		source_slot->effective_request_lsn = 0;
 
+		/* update bookkeeping */
 		n_moved++;
 	}
 
-	if (MyPState->ring_last != empty_ring_index)
+	/*
+	 * Only when we've moved slots we can expect trailing unused slots,
+	 * so only then we clean up trailing unused slots.
+	 */
+	if (n_moved > 0)
```

Comment:
Right now it is assumed that if `compact_prefetch_buffers` returns false than current slot is used.
But it is not quite clear to me how it is enforced with this fix.
There are two loos `while (search_ring_index > MyPState->ring_last)` in  `compact_prefetch_buffers` 
First one is finished if it locates unused slot.
As far as it decrements `search_ring_index` second loop can be executed 0 times. In this case `n_moved` will be zero a and  `compact_prefetch_buffers` returns false. But is their any warranty that slow pointed `MyPState->ring_last` is used?